Add osu!direct previews

jorolf opened the pull request.

jorolf committed:
add previews to osu!direct

jorolf committed:
Merge branch 'master' of git://github.com/ppy/osu into direct-previews

jorolf committed:
appveyor

jorolf committed:
Merge branch 'master' into direct-previews

peppy commented on osu.Game/Overlays/Direct/DirectPanel.cs:
```diff
@@ -38,6 +39,9 @@ public abstract class DirectPanel : Container
         private BeatmapManager beatmaps;
         private NotificationOverlay notifications;
 
+        public abstract Bindable<bool> PreviewPlaying { get; }
```
why not just make this a readonly non-abstract? not sure why it's abstract.

peppy commented on osu.Game/Overlays/Direct/DirectListPanel.cs:
```diff
@@ -128,7 +157,28 @@ private void load(LocalisationEngine localisation)
+                },
             });
+
+            PreviewPlaying.ValueChanged += newValue => playButton.FadeTo(newValue || IsHovered ? 1 : 0, 120, Easing.InOutQuint);
```
why is this not in the base implementation?

peppy commented on osu.Game/Overlays/Direct/DirectListPanel.cs:
```diff
@@ -128,7 +157,28 @@ private void load(LocalisationEngine localisation)
+            base.Update();
+
+            if (PreviewPlaying && playButton.Track != null)
+                progressBar.Width = (float)(playButton.Track.CurrentTime / playButton.Track.Length);
```
this is really ugly. at this point the track should be stored locally, not inside the button.

peppy commented:
Please check out the audio loading wrapper in https://github.com/ppy/osu/pull/1253 (which is now merged). You should be able to share a fair bit of implementation with it

peppy reviewed (changes requested):
as commented

jorolf committed:
make PreviewPlaying readonly instead of abstract

jorolf committed:
Merge branch 'master' of git://github.com/ppy/osu into direct-previews

jorolf committed:
move logic to DirectPanel and reuse stuff for the PreviewButton

jorolf committed:
Merge branch 'master' of git://github.com/ppy/osu into direct-previews

jorolf committed:
add license header

jorolf committed:
remove unused usings and reset the track to the start when playing again

jorolf committed:
another unused using

peppy committed:
Merge branch 'master' into direct-previews

peppy commented on osu.Game/Overlays/Direct/DirectPanel.cs:
```diff
@@ -102,12 +112,61 @@ private void load(APIAccess api, BeatmapManager beatmaps, OsuColour colours, Not
+        {
+            if (newValue)
+            {
+                if(Preview == null)
```
formatting miss (should be a space here)

peppy commented on osu.Game/Overlays/Direct/PlayButton.cs:
```diff
@@ -0,0 +1,90 @@
+            }
+        }
+
+        public PlayButton(Bindable<bool> playing)
```
should just be a public readonly bindable, not a ctor arg.

peppy commented:
Why didn't you include the actual preview loading logic inside `PlayButton`? ie. why do both PreviewButton and PlayButton exist, with the track logic duplicated in two places?

jorolf committed:
move stuff thats duplicated in PreviewButton and DirectPanel to PlayButton

jorolf committed:
Merge branch 'master' of git://github.com/ppy/osu into direct-previews

jorolf committed:
remove redundant stuff

peppy committed:
Apply some renames and refactoring of loading logic
Reduced publicly facing properties where possible. Also fixes a potentially bad state issue when the beatmapset was changed while a load was in progress.

peppy committed:
Merge remote-tracking branch 'upstream/master' into direct-previews

peppy committed:
Remove unused field
